Mathlib/Algebra/Polynomial/RuleOfSigns.lean

```diff
@@ -0,0 +1,588 @@
+    have hndeP0 : 0 < natDegree ((X - C η) * P.eraseLead) :=
+      Nat.lt_of_sub_eq_succ (hdp1 ▸ hndxeP)
+    rw [coeffList_eraseLead (ne_zero_of_natDegree_gt hndeP0)]
+    rw [List.map_cons, List.map_cons, leadingCoeff_eraseLead_eq_nextCoeff hQ₁.ne]
+    rw [sign_neg hQ₁, sign_neg ?_, this, hndxeP, hndexP0]
+    rwa [leadingCoeff_mul, leadingCoeff_X_sub_C, one_mul, leadingCoeff_eraseLead_eq_nextCoeff hc.ne]
+  rw [← self_sub_monomial_natDegree_leadingCoeff, leadingCoeff_eraseLead_eq_nextCoeff hQ₁.ne]
+  rw [hndexP0]
+  rw [← self_sub_monomial_natDegree_leadingCoeff, leadingCoeff_monic_mul (monic_X_sub_C η)]
+  rw [← self_sub_monomial_natDegree_leadingCoeff, leadingCoeff_monic_mul (monic_X_sub_C η)]
+  rw [hndxeP, hndxP]
+  rw [leadingCoeff_eraseLead_eq_nextCoeff hc.ne, ← self_sub_monomial_natDegree_leadingCoeff]
+  rw [hQ, mul_sub, sub_mul, sub_mul, X_mul_monomial, C_mul_monomial, monomial_sub]
+  rw [leadingCoeff, nextCoeff_of_natDegree_pos (hdp1 ▸ dp1.succ_pos), hdp1, Nat.add_sub_cancel]
+  abel
+
+--effectively a specialization of the final theorem for monomials
+lemma succ_sign_lin_mul_monomial {d c} (hc : c ≠ 0) (hη : η > 0) :
+    (monomial d c).SignVariations + 1 ≤ ((X - C η) * monomial d c).SignVariations := by
+  have h₁ : nextCoeff ((X - C η) * monomial d c) = -(η * c) := by
+    convert coeff_mul_monomial (X - C η) d 0 c using 1
+    · simp [hc, nextCoeff, natDegree_mul (X_sub_C_ne_zero η)]
+    · simp
+  have h₂ : eraseLead ((X - C η) * monomial d c) ≠ 0 := by
+    apply mt nextCoeff_eq_zero_of_eraseLead_eq_zero
+    simp [h₁, hc, hη.ne']
+  have h₃ : SignType.sign c ≠ SignType.sign (-(η * c)) := by
+    simp [hη, hc, Left.sign_neg, sign_mul]
+  simpa [h₁, h₂, h₃, hc, hη.ne', SignVariations, List.destutter_cons_cons,
+    ← leadingCoeff_cons_eraseLead, coeffList_eraseLead, leadingCoeff_eraseLead_eq_nextCoeff]
+  using List.length_pos_of_ne_nil (List.destutter'_ne_nil _ _)
+
+--Technically this is true as long as sign(P.leadingCoeff) = sign(P.nextCoeff), but we just
+--do the case where they're positive since that's easier (and sufficient for the downstream)
+lemma signvar_mul_eraseLead_le_of_nextCoeff (h₁ : leadingCoeff P > 0) (h₂ : P.nextCoeff > 0) :
+    SignVariations ((X - C η) * P.eraseLead) ≤ SignVariations ((X - C η) * P) := by
+  have hpne0 : P ≠ 0 :=
+    (leadingCoeff_ne_zero.mp h₁.ne')
+  have hndxp : natDegree ((X - C η) * P) = P.natDegree + 1 := by
+    rw [natDegree_mul (X_sub_C_ne_zero η) hpne0, natDegree_X_sub_C, add_comm]
+  have hndep : P.natDegree = P.eraseLead.natDegree + 1 :=
+    (natDegree_eraseLead_add_one h₂.ne').symm
+  have hxepn0 : (X - C η) * P.eraseLead ≠ 0 :=
+    mul_ne_zero (X_sub_C_ne_zero η) (mt nextCoeff_eq_zero_of_eraseLead_eq_zero h₂.ne')
+  have hndxep : natDegree ((X - C η) * eraseLead P) = P.eraseLead.natDegree + 1 := by
+    rw [natDegree_mul (mul_ne_zero_iff.mp hxepn0).1 (mul_ne_zero_iff.mp hxepn0).2,
+      natDegree_X_sub_C, add_comm]
+
+  have heqc : ∃ (c₀ : α) (cs : List α),
```

Comment:
Can this `have` be refactored into a lemma? Generally, it is nice if theorems with proofs more than a few dozen lines long are refactored to use lemmas as much as possible.